addons/mail/static/src/messaging/component/thread_viewer/thread_viewer.js

```diff
@@ -153,6 +153,10 @@ Object.assign(ThreadViewer, {
             type: Boolean,
             optional: true,
         },
+        hasComposerThreadTyping: {
+            type: Boolean,
+            optional: true,
```

Comment:
This is a prop to forward to composer component. Composer component defines a default for that prop, and we want to use such default if we do not forward a prop for that, hence prefer optional and consider `undefined` as if no such prop is given.

Reply:
I'm not sure if this is a good idea... `hasThreadTyping` on `Composer` is a technical detail that the one instantiating a `ThreadViewer` shouldn't be aware of. Having to check for the default value somewhere else is making it complicated and depends on implementation.

It is reasonable to assume we might just as well change the implementation to display this information somewhere else than the composer, for example at the bottom of the message list.

If we decide that `ThreadViewer` has the responsibility of knowing this value, it should have the responsibility of knowing its default value too.

Reply:
I don't really get how this justifies putting a default on thread viewer, in fact your comment encourages the contrary. `ThreadViewer` shouldn't indeed really care for `hasThreadTyping` of `Composer` if it didn't want to, hence the default values should be imposed by `Composer` itself. If you put a default on `ThreadViewer`, then you enforce that any use of `ThreadViewer` with a composer should always have (or not have) the typing stuff displayed on `Composer`, which goes against what you say and what we want (i.e. typing info on composer in Discuss thread viewer, but not in chat window thread viewer).

The general idea is to define defaults in component that really uses it (here `Composer`). Parented components may enforce a value (e.g. hard-coded value in template) or allow customization of sub-component by means of a prop on themselves (which corresponds exactly to `hasComposerThreadTyping `). This is a prop that has to be forwarded to `Composer`, and could be optional, which is the intend of this `optional` here. Putting a `default` here is even harmful, because this removes the possibility to say "I don't care"/"Just use composer with its default rendering".

**tl;dr** I strongly disagree with putting a default in `ThreadViewer`.

Reply:
I agree if we think it's the responsibility of the `Composer`, but it's not the case for me.

"Who is typing" is an information that belongs to a whole "thread" conceptually. It's a technical detail that we decided to display it on the composer area instead of elsewhere.

I would even go as far as saying it is indeed not the right place. It can be "below" the composer (visually), but it will always be inside the thread. I'm not composing anything, I might still want to know someone else is typing.

Reply:
I think you're mixing entities and components here. The notion of typing definitely belong to thread viewer entity, which is exactly how it is implemented. With components, however, the rendering of typing information is on the `Composer` component, so there is absolutely no reason to enforce everything on `ThreadViewer` (I would even say it's harmful, because you restrict having `ThreadViewer` (component) with and without thread typing on `Composer` (component) at your will on a case-by-case basis).

Also the prop `hasThreadTyping` **is not** a logic (entity-level) concern. It's just a UI customization (on/off) on a composer, thus this prop must be on the component that directly use it (with a default), and parented components may have a prop to forward this but they should strongly avoid enforcing a default: this is an anti-pattern (makes customization very complicated, bug-prone, overshadowing everywhere in a similar fashion to inheritance in OOP, etc.).

Reply:
Yes, I agree with the reasoning.
I think my only problem is the hypothesis.

I simply don't like that `ThreadTextualTypingStatus` is part of `Composer` component, instead of being part of `ThreadViewer` below the `Composer`.

Reply:
We want it below composer input, and above attachment list of composer, so we have to put in in `Composer`, we can't put it directly in template of `ThreadViewer`.

Reply:
Yep, basically I don't like the spec. We can merge it like this for now.

Reply:
Maybe just comment on the prop that we specifically allow undefined as a fallback to composer default.